cpp/src/compaction/stream_compaction_ops.cu

```diff
@@ -224,109 +230,232 @@ gdf_error gpu_apply_stencil(gdf_column *lhs, gdf_column * stencil, gdf_column *
 
 	size_t n_bytes = get_number_of_bytes_for_valid(stencil->size);
 
-	bit_position_iterator bit_position_iter(thrust::make_counting_iterator<gdf_size_type>(0), modulus_bit_width(n_bytes, stencil->size));
-	gdf_valid_iterator valid_iterator(thrust::detail::make_normal_iterator(thrust::device_pointer_cast(stencil->valid)),GDF_VALID_BITSIZE);
-	//TODO: can probably make this happen with some kind of iterator so it can work on any width size
-
-	//zip the stencil and the valid iterator together
-	typedef thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<int8_t> >,gdf_valid_iterator, bit_position_iterator > zipped_stencil_tuple;
-	typedef thrust::zip_iterator<zipped_stencil_tuple> zipped_stencil_iterator;
+	// If both valids are not nulls or if just one of them are nulls
+	if((col->valid != nullptr && stencil->valid != nullptr) || ((col->valid == nullptr || stencil->valid == nullptr) && col->valid != stencil->valid)) {
```

Comment:
This is a large chunk of code inside this `if` case. Consider wrapping it in a function.

Reply:
Refactored code.